### ISSUE TITLE
Replace the palette-packing algorithm with the one used by Superfamiconv?

### ISSUE THREAD

**Issue by Rangi42** [gbdev/rgbds#1451]
Discussion #1427 notes that the current palette-packing algorithm, based on a paper on bin packing, fails in some cases where a packing should be possible. The algorithm used by superfamiconv (and ported to Tilemap Studio) works for that test case. Unless it has some other flaw, we should probably switch.

**Comment by ISSOtm** [comment]
I think we shouldn't *replace* the algorithm. Since we already have the more complex one, I think we should keep both, run both of them (they fast so that should be okay?), and keep the best results.

I think the naive packer is fairly easy to “get to fail”; the question then becomes how common either case is.

But if we keep both, we can have the best of both worlds—at a slight speed loss, but hey, if someone's concerned about that, we *could* provide an option to only run one or the other...

**Comment by Rangi42** [comment]
I have not yet seen the "naive" packer fail; it's done well in superfamiconv for years.

I really don't want to maintain two separate pal-packing strategies, just on the off-chance that one may fail (especially when it's the more complex code which has been observed to fail).

**Comment by ISSOtm** [comment]
To be fair, the failure wouldn't have been brought to our attention if I hadn't been the first person to encounter it *and* I hadn't obtained the image from `-r` in the first place—I'd have assumed that the packing was impossible.

I don't feel confident that this scenario hasn't happened to SuperFamiconv's algorithm either.

**Comment by Rangi42** [comment]
I could try to improve the current packing algo, but it's complex enough (in implementation if not in underlying logic) that I'm not certain of doing that. If the superfamiconv one failed, it would probably be easier to patch.

That said, if you can fix it for this one known failure, I'd be fine with closing the issue.

**Comment by aaaaaa123456789** [comment]
As far as I know, this problem is NP-hard. In other words, there's no way to write a single packer that will work for all cases and won't run unacceptably slow quite often.
If you have two packers that cover most common cases, then running both is certainly a reasonable alternative. (You can even run them in parallel if you wish.) The "correct" solution of using a single packer that handles all cases is simply not reachable, unless you feel like proving P = NP.

**Comment by Rangi42** [comment]
It's NP-hard for the general case. We're fitting items into bins of size 4, with usually not more than 8 bins total (and never more than 256). This is the same scale where people still use bubble-sort because "short simple code that runs fast enough" works.

I would bet that after basic preprocessing (removing duplicates and subsets), we could literally iterate through all possible arrangements and still not be significantly slower. I also have yet to see anyone give an example of the superfamiconv algorithm failing to pack a palette, and I expect that at this problem scale, its approach is not actually beat by this Overload-and-Remove heuristic from the paper.

Basically, at our scale, the problem is small enough that I'm really averse to maintaining such a large chunk of code around to deal with it.

**Comment by ISSOtm** [comment]
Here is an example of an image that SuperFamiconv fails to pack correctly:

![Apparently random stripes of colour](https://github.com/user-attachments/assets/97ec514a-9eb9-475c-8651-e52544baa8bf)

RGBGFX v0.8.0 correctly packs it as
```
{ 001f, 03f0, 7fe0, 7c10, }
{ 02ff, 23e0, 7d00, 5c1f, }
```
but SuperFamiconv instead generates:
```json
[["#ff00bd","#00ffff","#0042ff","#8400ff"],
 ["#84ff00","#00ff42","#8400ff","#ff00bd"],
 ["#84ff00","#00ff42","#00ffff","#0042ff"],
 ["#ff0000","#ffbd00","#8400ff","#ff00bd"],
 ["#ff0000","#ffbd00","#00ffff","#0042ff"],
 ["#ff0000","#ffbd00","#84ff00","#00ff42"]]
```

(Image generated using [this little doodad](https://github.com/ISSOtm/no-packsaran).)

**Comment by Rangi42** [comment]
Nice, thank you! This can be closed then.
